proxylib/memcached/text/parser.go

```diff
@@ -0,0 +1,331 @@
+				return proxylib.ERROR, 0
+			}
+			// 2 additional bytes for terminating linefeed
+			frameLength = linefeed + 2 + nBytes + 2
+
+			if command[0] == 'c' { //storage command is "cas"
+				noreply = len(tokens) == CAS_WITH_NOREPLY_FIELDS
+			} else {
+				noreply = len(tokens) == STORAGE_WITH_NOREPLY_FIELDS
+			}
+		case p.isCommandDelete(command):
+			meta.Keys = tokens[1:2]
+			noreply = len(tokens) == DELETE_WITH_NOREPLY_FIELDS
+			frameLength = linefeed + 2
+		case p.isCommandIncrDecr(command):
+			meta.Keys = tokens[1:2]
+			noreply = len(tokens) == INCR_WITH_NOREPLY_FIELDS
+			frameLength = linefeed + 2
+		case bytes.Equal(command, []byte("touch")):
+			meta.Keys = tokens[1:2]
+			noreply = len(tokens) == TOUCH_WITH_NOREPLY_FIELDS
+			frameLength = linefeed + 2
+		case bytes.Equal(command, []byte("slabs")) ||
+			bytes.Equal(command, []byte("lru")) ||
+			bytes.Equal(command, []byte("lru_crawler")) ||
+			bytes.Equal(command, []byte("stats")) ||
+			bytes.Equal(command, []byte("version")) ||
+			bytes.Equal(command, []byte("misbehave")):
+
+			meta.Keys = [][]byte{}
+			frameLength = linefeed + 2
+		case bytes.Equal(command, []byte("flush_all")) ||
+			bytes.Equal(command, []byte("cache_memlimit")):
+			meta.Keys = [][]byte{}
+			noreply = bytes.Equal(tokens[len(tokens)-1], []byte("noreply"))
+			frameLength = linefeed + 2
+		case bytes.Equal(command, []byte("quit")):
+			meta.Keys = [][]byte{}
+			noreply = true
+			frameLength = linefeed + 2
+		case bytes.Equal(command, []byte("watch")):
+			meta.Keys = [][]byte{}
+			frameLength = linefeed + 2
+			p.watching = true
+		default:
+			log.Error("Could not parse text memcache frame")
+			return proxylib.ERROR, 0
+		}
+		logEntry := &cilium.LogEntry_GenericL7{
+			&cilium.L7LogEntry{
+				Proto: "textmemcached",
+				Fields: map[string]string{
+					"command": meta.Command,
+					"keys":    string(bytes.Join(meta.Keys, []byte(", "))),
+				},
+			},
+		}
+
+		p.requestCount++
+		r := &replyIntent{
+			requestID: p.requestCount,
+			command:   command,
+		}
+
+		matches := p.connection.Matches(meta)
+
+		if matches {
+			r.denied = false
+			p.replyQueue = append(p.replyQueue, r)
+			p.connection.Log(cilium.EntryType_Request, logEntry)
+			return proxylib.PASS, frameLength
+		}
+
+		r.denied = true
+		if !noreply {
+			p.injectDeniedMessage()
+			if p.requestCount == p.replyCount+1 {
+				p.injectDeniedMessage()
+			} else {
+				p.replyQueue = append(p.replyQueue, r)
+			}
+		}
+		p.connection.Log(cilium.EntryType_Denied, logEntry)
+		return proxylib.DROP, frameLength
+	}
+
+	//reply
+	log.Debugf("reply, parsing to figure out if we have it all")
+
+	intent := p.replyQueue[0]
+
+	logEntry := &cilium.LogEntry_GenericL7{
+		&cilium.L7LogEntry{
+			Proto: "textmemcached",
+			Fields: map[string]string{
+				"command": string(intent.command),
+			},
+		},
+	}
+	if p.watching {
+		// in watch mode we pass all replied lines
+		return proxylib.PASS, linefeed + 2
```

Comment:
Why are we not popping the `replyQueue` in this case?

Reply:
Is it not possible to detect the actual frame size of the reply so that we could reliably pair replies with requests and correctly pop the `replyQueue` and consequently issue the denied messages as queued, and also access log the replies? Or maybe I misunderstand the purpose of the "watch" mode?

Reply:
Basically if connection is in watch mode, no further commands will be accepted by server, and it will provide client with watch lines until connection closes.

Excerpt from docs:
```
watch <fetchers|mutations|evictions>

- Turn connection into a watcher. Options can be stacked and are
  space-separated. Logs will be sent to the watcher until it disconnects.
```

https://github.com/memcached/memcached/blob/master/doc/protocol.txt#L578